src/sacn/source.c

```diff
@@ -595,13 +1232,695 @@ etcpal_error_t sacn_source_reset_networking(sacn_source_t handle, uint16_t unive
  * @return The total number of network interfaces for the universe. If this is greater than netints_size, then only
  * netints_size addresses were written to the netints array. If the source or universe were not found, 0 is returned.
  */
-size_t sacn_source_get_network_interfaces(sacn_source_t handle, uint16_t universe, SacnMcastInterface* netints,
+size_t sacn_source_get_network_interfaces(sacn_source_t handle, uint16_t universe, EtcPalMcastNetintId* netints,
                                           size_t netints_size)
 {
-  ETCPAL_UNUSED_ARG(handle);
-  ETCPAL_UNUSED_ARG(universe);
-  ETCPAL_UNUSED_ARG(netints);
-  ETCPAL_UNUSED_ARG(netints_size);
+  size_t total_num_network_interfaces = 0;
+
+#if SACN_SOURCE_ENABLED
+  if (sacn_lock())
+  {
+    // Look up universe state
+    SacnSourceUniverse* universe_state = NULL;
+    if (lookup_source_state(handle, universe, NULL, &universe_state) == kEtcPalErrOk)
+    {
+      total_num_network_interfaces = universe_state->num_netints;
+
+      // Copy out the netints
+      for (size_t i = 0; netints && (i < netints_size) && (i < total_num_network_interfaces); ++i)
+        netints[i] = universe_state->netints[i];
+    }
+
+    sacn_unlock();
+  }
+#endif
+
+  return total_num_network_interfaces;
+}
+
+bool source_handle_in_use(int handle_val, void* cookie)
+{
+  ETCPAL_UNUSED_ARG(cookie);
+
+  return (handle_val == SACN_SOURCE_INVALID) || (lookup_source_state(handle_val, 0, NULL, NULL) == kEtcPalErrOk);
+}
+
+// Needs lock
+etcpal_error_t start_tick_thread()
+{
+  shutting_down = false;
+  EtcPalThreadParams params = ETCPAL_THREAD_PARAMS_INIT;
+
+  if (etcpal_thread_create(&source_thread_handle, &params, source_thread_function, NULL) != kEtcPalErrOk)
+    return kEtcPalErrSys;
+
+  return kEtcPalErrOk;
+}
+
+// Takes lock
+void stop_tick_thread()
+{
+  etcpal_thread_t thread_handle;
+
+  if (sacn_lock())
+  {
+    shutting_down = true;  // Trigger thread-based sources to terminate
+    thread_handle = source_thread_handle;
+    sacn_unlock();
+  }
+
+  // Wait for thread-based sources to terminate (assuming application already cleaned up manual sources)
+  etcpal_thread_join(&thread_handle);
+}
+
+// Takes lock
+void source_thread_function(void* arg)
+{
+  ETCPAL_UNUSED_ARG(arg);
+
+  bool keep_running_thread = true;
+  int num_thread_based_sources = 0;
+
+  EtcPalTimer interval_timer;
+  etcpal_timer_start(&interval_timer, SOURCE_THREAD_INTERVAL);
+
+  while (keep_running_thread || (num_thread_based_sources > 0))
+  {
+    num_thread_based_sources = process_internal(false);
+
+    etcpal_thread_sleep(etcpal_timer_remaining(&interval_timer));
+    etcpal_timer_reset(&interval_timer);
+
+    if (sacn_lock())
+    {
+      keep_running_thread = !shutting_down;
+      sacn_unlock();
+    }
+  }
+}
+
+// Takes lock
+int process_internal(bool process_manual)
```

Comment:
This function feels like it could be better named, process internal what?

Reply:
This function is pretty much process_sources + locking. It could either be a name that indicates that, or I could get rid of this function and just do the locking within the two functions that use it.

Reply:
take_lock_and_process_sources?

Reply:
Resolved in https://github.com/ETCLabs/sACN/pull/32/commits/fab065959ef9532213c8e2d6819f30fea1f929c5.